Documentation/coding-guidelines/breaking-change-rules.md

```diff
@@ -185,8 +186,17 @@ Breaking Change Rules
 
 * Change from `ref readonly` return to `ref` return (except for virtual methods or interfaces)
 
+* Adding an interface method with a default implementation to an interface
+
+    Note that care must be taken when adding a default implementation that has "update" semantic (e.g. adding `void AddAll(IEnumerable<T> items)` to `ICollection<T>`). If the interface is implementated by a struct, the default implementation is executed on a boxed `this`. Changes to state done by the default method might be lost.
```

Comment:
> If the interface is implementated by a struct, the default implementation is executed on a boxed `this`

This is true even if the calls are already being made on a box? e.g.
```C#
ICollection<T> c = someStruct;
c.AddAll(...);
```
?

This seems like a real pit of failure; essentially one should never implement an interface with default members on a struct, and never add default members to an interface that might be implemented on a struct.

Reply:
If user boxes on purpose (e.g. they cast to the interface), all calls happen on the boxed version - default implementations are no different from implementations on the struct in that case.

This is only a problem in situations where we have a constrained call and the runtime does the boxing on user's behalf:

```csharp
void Do<T>(ref T x) where T : ICollection<T>
{
    // No boxing here, constrained call resolved to a direct call to an instance method on the struct at compile time
    x.Add(...);

    // Secret boxing here
    x.AddAll(...);
}
```

Reply:
Ok, phew.  That's still not great, but it's a lot less not great :wink:

Reply:
Can we clarify that in the text you added?

Reply:
How about this?

```suggestion
    Note that care must be taken when adding a default implementation that has "update" semantic (e.g. adding `void AddAll(IEnumerable<T> items)` to `ICollection<T>`). If the interface is implemented by a struct, the default implementation is always executed on a boxed `this`. If the struct is not boxed, the runtime boxes it on users behalf. Changes done by the default interface method would be lost in that case. An example where this implicit boxing would happen are constrained calls (`void CallMethod<T, U>(ref T x) where T : struct, ICollection<U> { x.AddAll(...); }`
```

Reply:
Sounds good, thanks.